streaming/base/partitioning.py

```diff
@@ -29,71 +29,76 @@ def get_partitions(num_samples: int,
         drop_first (int): Number of samples seen already, which are dropped. Defaults to ``0``.
 
     Returns:
-        NDArray[np.int64]: Partitions of shape (physical nodes x ranks x workers x samples).
+        NDArray[np.int64]: Partitions (physical nodes x ranks x workers x batches x samples).
     """
-    # Divide the full dataset sample range into sample range per canonical node.
     batch_size = batch_size or 1
-    xx = []
-    for i in range(num_canonical_nodes):
-        a = i * num_samples // num_canonical_nodes
-        z = (i + 1) * num_samples // num_canonical_nodes
-        x = np.arange(a, z)
-        xx.append(x)
 
-    # Make the spans equal length by repeating the last sample if too short.
-    # -> Shape: (canonical nodes x samples).
-    max_len = max(map(len, xx))
-    for i, x in enumerate(xx):
-        if len(x) < max_len:
-            last = np.array([x[-1]])
-            xx[i] = np.concatenate([x, last])
-    x = np.stack(xx)
-
-    # If there are more physical than canonical nodes, interleave canonical onto physical nodes.
-    # -> Shape: (canonical nodes x samples).
+    # Divide the full dataset sample range into a sample range per canonical node.
+    samples_per_canonical_node = (num_samples + num_canonical_nodes - 1) // num_canonical_nodes
+    node_ratio = 0
+    padding = 0
     if num_canonical_nodes < num_physical_nodes:
         assert not num_physical_nodes % num_canonical_nodes
-        ratio = num_physical_nodes // num_canonical_nodes
-        too_many = x.shape[1] % ratio
-        if too_many:
-            too_few = ratio - too_many
-            last = x[:, -ratio - too_few + 1:-ratio + 1]
-            x = np.concatenate([x, last], 1)
+        node_ratio = num_physical_nodes // num_canonical_nodes
+        overflow = samples_per_canonical_node % node_ratio
+        if overflow:
+            padding = node_ratio - overflow
+    padded_samples_per_canonical_node = samples_per_canonical_node + padding
+
+    # Create the initial sample ID matrix.
+    #
+    # Shape: (canonical nodes, padded samples per canonical node).
+    x = np.arange(num_canonical_nodes * padded_samples_per_canonical_node, dtype=np.int64)
+    x = x.reshape(num_canonical_nodes, padded_samples_per_canonical_node)
 
-    # Drop samples that have already been seen and reshape.
-    # -> Shape: (physical nodes x samples).
+    # Make adustments to replicate the original padding and extending behavior.
+    offsets = np.arange(num_canonical_nodes) * padding
+    offsets = np.expand_dims(offsets, 1)
```

Comment:
What would be shape after `offsets = np.expand_dims(offsets, 1)` ? is it `Shape: (canonical nodes, padded samples per canonical node)` ? Can you please add the shape as a comment to understand it better in future ?

Reply:
(C, 1)